### PR TITLE
Introduce wildcard context bounds

### PR DESCRIPTION
**Pull request by adampauls**
Implements the proposal discussed [here](https://contributors.scala-lang.org/t/pre-sip-additional-syntactic-sugar-for-typeclasses/5675/27): allows

```scala
def showAll(xs: List[? : Showable]): Unit = xs.foreach(x => println(x.show))
```
to be sugar for 
```scala
def showAll[S: Showable](xs: List[S]): Unit = xs.foreach(x => println(x.show))
``` 
which is already sugar for
```scala
def showAll[S](xs: List[S])(using Showable[S]): Unit = xs.foreach(x => println(x.show))
```

As far as I understand, the SIP process is dormant right now, so I'm not sure how to get agreement on these things. I thought I'd put up the PR to show how simple this change is, in the likely-unrealistic hopes that it makes it in before 3.2.0.

### CODE REVIEW

**Review by odersky — CHANGES_REQUESTED** [review]
I think this is an excellent change overall, but we have to wait on this until the SIP process starts up again. 

I have not yet reviewed the code in detail, but one thing that's missing is a doc page that explains and specs the new feature.

**Comment by adampauls** [comment]
Thanks, I'll figure out how to add the import. Is a `-Y` flag also an option, or are those dispreferred now?

**Comment by adampauls** [comment]
(Ah, I see you get both for free, nevermind).

**Comment by adampauls** [comment]
I hid it behind an experiment feature flag/import and added a documentation page.

**Comment by adampauls** [comment]
Friendly bump.

**Comment by som-snytt** [comment]
Fist bump. 🤜 

Alternative syntax for anonymous context bound:
```
def showAll(xs: List[🕶️ : Showable]): Unit
```

File: docs/_docs/reference/experimental/anonymous-context-bounds.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+```
+
+is sugar for
+
+```scala
+def showAll[S: Showable](xs: Set[S]): Unit = xs.foreach(x => println(x.show))
+```
+
+which further desugars as above. Rust, which also has first-class support for typeclasses,
+has a similar syntax with its [`impl` traits](https://doc.rust-lang.org/reference/types/impl-trait.html)
```
Comment (bishabosha) [inline docs/_docs/reference/experimental/anonymous-context-bounds.md:46]: Swift has introduced a very similar idea with its "opaque parameter declarations"
https://github.com/apple/swift-evolution/blob/main/proposals/0341-opaque-parameters.md

TL;DR - `func showAll<A: Showable>(_: [A]) -> Void` => `func showAll(_: [some Showable]) -> Void`

**Review by bishabosha — CHANGES_REQUESTED** [review]
I found it to crash with tuple types: (I used repl)
```scala
import language.experimental.anonymousContextBounds

trait Showable[-A]:
  extension (a: A) def show: String

def showPair(p: (? : Showable, ? : Showable)): String = ???
```

<details>

<summary>see stack trace</summary>

```
[error] scala.MatchError: ContextBounds(TypeBoundsTree(EmptyTree,EmptyTree,EmptyTree),List(AppliedTypeTree(Ident(Showable),List(Ident(?))))) (of class dotty.tools.dotc.ast.untpd$ContextBounds)
[error] 	at dotty.tools.dotc.ast.desugar$.apply(Desugar.scala:1835)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2903)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2919)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2984)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2988)
[error] 	at dotty.tools.dotc.typer.Typer.typedType(Typer.scala:3107)
[error] 	at dotty.tools.dotc.typer.Typer.typedArg$1(Typer.scala:2000)
[error] 	at dotty.tools.dotc.typer.Typer.$anonfun$37(Typer.scala:2004)
[error] 	at dotty.tools.dotc.core.Decorators$ListDecorator$.zipWithConserve$extension(Decorators.scala:151)
[error] 	at dotty.tools.dotc.typer.Typer.typedAppliedTypeTree(Typer.scala:2004)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2882)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2919)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2984)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2988)
[error] 	at dotty.tools.dotc.typer.Typer.typedTuple(Typer.scala:2766)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2893)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2919)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2984)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2988)
[error] 	at dotty.tools.dotc.typer.Typer.typedType(Typer.scala:3107)
[error] 	at dotty.tools.dotc.typer.Namer.typedAheadType$$anonfun$1(Namer.scala:1618)
[error] 	at dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1611)
[error] 	at dotty.tools.dotc.typer.Namer.typedAheadType(Namer.scala:1618)
[error] 	at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1685)
[error] 	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:782)
[error] 	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:921
[error] 	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:809)
[error] 	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:167)
[error] 	at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:187)
[error] 	at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:189)
[error] 	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:369)
[error] 	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2797)
[error] 	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2822)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2918)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2984)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2988)
[error] 	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3104)
[error] 	at dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1621)
[error] 	at dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1611)
[error] 	at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1621)
[error] 	at dotty.tools.dotc.typer.Namer.completeParams$$anonfun$1(Namer.scala:1636)
[error] 	at scala.collection.immutable.List.foreach(List.scala:333)
[error] 	at dotty.tools.dotc.typer.Namer.completeParams(Namer.scala:1636)
[error] 	at dotty.tools.dotc.typer.Namer.defDefSig$$anonfun$3(Namer.scala:1725)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] 	at scala.collection.immutable.List.foreach(List.scala:333)
[error] 	at dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1725)
[error] 	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:786)
[error] 	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:921
[error] 	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:809)
[error] 	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:167)
[error] 	at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:187)
[error] 	at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:189)
[error] 	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:369)
[error] 	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2797)
[error] 	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2822)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2918)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2984)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2988)
[error] 	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3010)
[error] 	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3060)
[error] 	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2498)
[error] 	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:2844)
[error] 	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2848)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2918)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2984)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2988)
[error] 	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3010)
[error] 	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3060)
[error] 	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2627)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2889)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2919)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2984)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2988)
[error] 	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3104)
[error] 	at dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$1(TyperPhase.scala:43)
[error] 	at dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$adapted$1(TyperPhase.scala:50)
[error] 	at scala.Function0.apply$mcV$sp(Function0.scala:39)
[error] 	at dotty.tools.dotc.core.Phases$Phase.monitor(Phases.scala:414)
[error] 	at dotty.tools.dotc.typer.TyperPhase.typeCheck(TyperPhase.scala:50)
[error] 	at dotty.tools.dotc.typer.TyperPhase.runOn$$anonfun$3(TyperPhase.scala:84)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] 	at scala.collection.immutable.List.foreach(List.scala:333)
[error] 	at dotty.tools.dotc.typer.TyperPhase.runOn(TyperPhase.scala:84)
[error] 	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:225)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] 	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1328)
[error] 	at dotty.tools.dotc.Run.runPhases$1(Run.scala:236)
[error] 	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:244)
[error] 	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:253)
[error] 	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
[error] 	at dotty.tools.dotc.Run.compileUnits(Run.scala:253)
[error] 	at dotty.tools.dotc.Run.compileUnits(Run.scala:192)
[error] 	at dotty.tools.repl.ReplCompiler.runCompilationUnit(ReplCompiler.scala:166)
[error] 	at dotty.tools.repl.ReplCompiler.compile(ReplCompiler.scala:177)
[error] 	at dotty.tools.repl.ReplDriver.compile(ReplDriver.scala:287)
[error] 	at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:249)
[error] 	at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:168)
[error] 	at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:171)
[error] 	at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:192)
[error] 	at dotty.tools.repl.ReplDriver.runBody$$anonfun$1(ReplDriver.scala:180)
[error] 	at dotty.tools.runner.ScalaClassLoader$.asContext(ScalaClassLoader.scala:80)
[error] 	at dotty.tools.repl.ReplDriver.runBody(ReplDriver.scala:180)
[error] 	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:171)
[error] 	at xsbt.ConsoleInterface.run(ConsoleInterface.java:52)
```

</details>

File: docs/_docs/reference/experimental/anonymous-context-bounds.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+and
+
+```scala
+def showAll[S <: Showable2](xs: Set[S]): Unit = xs.foreach(x => println(x.show))
+```
+
+in the case where `Showable2` is a simple trait instead of one designed to be used as a typeclass.
+
+Unlike anonymous type bounds, anonymous context bounds can appear as the top-level type of a parameter
```
Comment (bishabosha) [inline docs/_docs/reference/experimental/anonymous-context-bounds.md:64]: I don't think it is widely used terminology to use "anonymous type bounds"
```suggestion
Unlike wildcard type arguments, anonymous context bounds can appear as the top-level type of a parameter
```

File: docs/_docs/reference/experimental/anonymous-context-bounds.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+```scala
+def showOne(x: ? : Showable): Unit = x.show // okay
+def showOne[S <: Showable2](x: S): Unit = x.show // okay
+def showOne(x: ? <: Showable2): Unit = x.show // error: Unbound wildcard type
```
Comment (bishabosha) [inline docs/_docs/reference/experimental/anonymous-context-bounds.md:69]: just a suggestion but could `Ordered` and `Ordering` from the standard library be the better comparison for when you would use subtyping vs context bound?

File: docs/_docs/reference/experimental/anonymous-context-bounds.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+The reason for publishing this extension now is to get feedback on its usefulness.
+
+## Why Anonymous Context Bounds?
+
+Typeclasses are [first-class consideration of Scala 3](https://docs.scala-lang.org/scala3/book/ca-type-classes.html) 
+and [context bound sugar](https://docs.scala-lang.org/scala3/book/ca-context-bounds.html) already exists to support their use.
+Taking an example from [here](https://docs.scala-lang.org/scala3/book/ca-type-classes.html), context bounds allow one
+to write
+```scala
+def showAll[S: Showable](xs: Set[S]): Unit = xs.foreach(x => println(x.show))
```
Comment (bishabosha) [inline docs/_docs/reference/experimental/anonymous-context-bounds.md:19]: for clarity, you should show the definition of `Showable` first - but I made a suggestion lower in the file that perhaps `Ordering` is fine enough and can link to the definition

File: docs/_docs/reference/experimental/anonymous-context-bounds.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+Unlike anonymous type bounds, anonymous context bounds can appear as the top-level type of a parameter
+
+```scala
+def showOne(x: ? : Showable): Unit = x.show // okay
+def showOne[S <: Showable2](x: S): Unit = x.show // okay
+def showOne(x: ? <: Showable2): Unit = x.show // error: Unbound wildcard type
+```
+
+This is because `def showOne(x: ? <: Showable2)` is not meaningfully different from `def showOne(x: Showable2)`,
```
Comment (bishabosha) [inline docs/_docs/reference/experimental/anonymous-context-bounds.md:72]: Im not sure this analogy works, because if `def showOne(x: ? <: Showable2)` meant the same as `def showOne[S <: Showable2](x: S): S` then potentially we get a more refined result type than just Showable2

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 6]
```diff
@@ -208,13 +216,19 @@ object desugar {
...
     }
 
-  /** 1. Expand context bounds to evidence params. E.g.,
+  /** 1. Expand anonymous bounds type params (if Feature.anonymousContextBounds enabled), E.g,
+    *
+    *      def f(params: ? : B)
+    *  ==>
+    *      def f[$_0 : B](params)
+    *
+    * 2. Expand context bounds to evidence params. E.g.,
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala:225]: margin is one space too much

File: docs/_docs/reference/experimental/anonymous-context-bounds.md (hunk 2 above)
Comment (adampauls) [reply docs/_docs/reference/experimental/anonymous-context-bounds.md:64]: Thanks. I switched all terminology from "anonymous [context|type] bound" to "wildcard [context | type] bounds"

As a nit: are they "wildcard type arguments", or just "wildcard types"? I think for them to be "arguments" (or "parameters"), they would need to appear in in the `[]` list?

File: docs/_docs/reference/experimental/anonymous-context-bounds.md (hunk 5 above)
Comment (adampauls) [reply docs/_docs/reference/experimental/anonymous-context-bounds.md:72]: Okay, but why is it disallowed then? Should I just leave that as a mystery?

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 6 above)
Comment (adampauls) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala:225]: Thanks, good catch.

File: docs/_docs/reference/experimental/anonymous-context-bounds.md (hunk 3 above)
Comment (adampauls) [reply docs/_docs/reference/experimental/anonymous-context-bounds.md]: I think Ordered/Ordering is confusing because they both take a type parameter. I will use ShowableTypeclass and ShowableInheritance

File: docs/_docs/reference/experimental/anonymous-context-bounds.md (hunk 4 above)
Comment (adampauls) [reply docs/_docs/reference/experimental/anonymous-context-bounds.md]: Done

File: docs/_docs/reference/experimental/anonymous-context-bounds.md (hunk 5 above)
Comment (adampauls) [reply docs/_docs/reference/experimental/anonymous-context-bounds.md]: I just removed the sentence completely.

**Comment by adampauls** [comment]
I realized there are a few more cases I need to cover, so this is not quite ready to merge yet.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 7]
```diff
@@ -1444,21 +1447,26 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
...
-                .orElse(errorType(AnonymousFunctionMissingParamType(param, tree, formal), param.srcPos))
-            val paramTpt = untpd.TypedSplice(
-                (if knownFormal then InferredTypeTree() else untpd.TypeTree())
-                  .withType(paramType.translateFromRepeated(toArray = false))
-                  .withSpan(param.span.endPos)
-              )
-            cpy.ValDef(param)(tpt = paramTpt)
+        for ((param, i) <- params.zipWithIndex) yield {
+          param.tpt match
+            case tree: untpd.ContextBounds =>
```
Comment (adampauls) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:1452]: The diff looks big, but only this new case was added.

**Review by odersky — COMMENTED** [review]
I am quite hopeful for this PR. Now that we have the SIP process revived, I think the best way forward is to submit this as a SIP. The discussion on contributors has been quiet for a while. I think the last comment https://contributors.scala-lang.org/t/pre-sip-additional-syntactic-sugar-for-typeclasses/5675/37 warrants a discussion, though.

After that the step would be the following:

 - submit this as a SIP
 - if the initial reviews are positive, submit the implementation as an experimental branch (i.e. revisit this PR)
 - once the SIP is accepted merge the implementation at the next minor language bump.

**Comment by adampauls** [comment]
After discussion on a [SIP](https://github.com/scala/improvement-proposals/pull/55), I've decided to abandon this feature.
